examples/training/classification/densenet/imagenet/basic_training.py

```diff
@@ -0,0 +1,224 @@
+    "weights_path", None, "Directory which will be used to store weight checkpoints."
+)
+
+FLAGS = flags.FLAGS
+FLAGS(sys.argv)
+
+NUM_CLASSES = 1000
+BATCH_SIZE = 256
+IMAGE_SIZE = (300, 300)
+EPOCHS = 250
+
+"""
+## Data loading
+This guide uses the
+[Imagenet dataset](https://www.tensorflow.org/datasets/catalog/imagenet2012). Note that this requires manual download, and does not work out-of-the-box with TFDS.
+To get started, we first load the dataset from a command-line specified directory where ImageNet is stored as TFRecords.
+"""
+
+
+def parse_imagenet_example(example, IMAGE_SIZE):
+    # Read example
+    image_key = "image/encoded"
+    label_key = "image/class/label"
+    keys_to_features = {
+        image_key: tf.io.FixedLenFeature((), tf.string, ""),
+        label_key: tf.io.FixedLenFeature([], tf.int64, -1),
+    }
+    parsed = tf.io.parse_single_example(example, keys_to_features)
+
+    # Decode and resize image
+    image_bytes = tf.reshape(parsed[image_key], shape=[])
+    image = tf.io.decode_jpeg(image_bytes, channels=3)
+    image = tf.image.resize(image, IMAGE_SIZE)
+
+    # Decode label
+    label = tf.cast(tf.reshape(parsed[label_key], shape=()), dtype=tf.int32) - 1
+    label = tf.one_hot(label, NUM_CLASSES)
+    return image, label
+
+
+def load_imagenet_dataset():
+    train_filenames = [
+        f"{FLAGS.imagenet_path}/train-{i:05d}-of-01024" for i in range(0, 1024)
+    ]
+    validation_filenames = [
+        f"{FLAGS.imagenet_path}/validation-{i:05d}-of-00128" for i in range(0, 128)
+    ]
+
+    train_dataset = tf.data.TFRecordDataset(filenames=train_filenames)
+    validation_dataset = tf.data.TFRecordDataset(filenames=validation_filenames)
+
+    train_dataset = train_dataset.map(
+        lambda x: parse_imagenet_example(x, IMAGE_SIZE),
+        num_parallel_calls=tf.data.AUTOTUNE,
+    )
+    validation_dataset = validation_dataset.map(
+        lambda x: parse_imagenet_example(x, IMAGE_SIZE),
+        num_parallel_calls=tf.data.AUTOTUNE,
+    )
+
+    return train_dataset.batch(BATCH_SIZE), validation_dataset.batch(BATCH_SIZE)
+
+
+train, test = load_imagenet_dataset()
+
+
+"""
+Next, we augment our dataset. We define a set of augmentation layers and then apply them to our input dataset using the `apply_augmentation` method from our KerasCV training utils.
+"""
+
+AUGMENT_LAYERS = [
+    keras_cv.layers.RandomFlip(),
+    keras_cv.layers.RandAugment(value_range=(0, 255), magnitude=0.3),
+    keras_cv.layers.RandomCutout(height_factor=0.1, width_factor=0.1),
+]
+
+
+@tf.function
+def augment(img, label):
+    inputs = {"images": img, "labels": label}
+    for layer in AUGMENT_LAYERS:
+        inputs = layer(inputs)
+    return inputs["images"], inputs["labels"]
+
+
+train = train.map(augment, num_parallel_calls=tf.data.AUTOTUNE).prefetch(
+    tf.data.AUTOTUNE
+)
+
+
+"""
+Now we can begin training our model. We begin by loading a DenseNet model from KerasCV.
+"""
+
+
+def get_model():
```

Comment:
lets consider inlining this

Reply:
I find this a bit more coherent with the rest of the script, especially for future expansions which may include pretraining. Keeping as much config as possible out of the main training loop will make it easier to follow, and isolate model-specific parameters to this section of the script.

What are your thoughts?

Reply:
Hmm, good point.  I think I'd drop all of these methods and prefer:

``
model = DenseNet121(
        include_rescaling=True,
        include_top=True,
        num_classes=NUM_CLASSES,
        input_shape=IMAGE_SIZE + (3,),
)
optimizer = optimizers.Adam(
        learning_rate=optimizers.schedules.PolynomialDecay(
            initial_learning_rate=0.005,
            decay_steps=train_ds.cardinality().numpy() * EPOCHS / 2,
            end_learning_rate=0.0001,
        )
    )
model.compile(
    optimizer=optimizer, 
    loss=losses.CategoricalCrossEntropy(label_smoothing=0.1), 
    metrics=['accuracy'],
    callbacks=get_callbacks()
)
```

in terms of readability.   My reasoning is that all of the model configuration and training is localized, but the callbacks (which are probably the least important) are non-localized.

That being said...  I dont have a particularly strong opinion and the line I am drawing is 100% arbitrary so I'm open to anything here.

Reply:
@LukeWood @ianjjohnson 

I believe the usage varies from case to case and in this particular case Luke's suggestion feels natural to me.
 
Personally, I prefer using the other approach (involving many `get_` functions) in research-like scenarios where the model, loss and optimizer might change frequently and so the code needs to be flexible (as a result of being modular).

Reply:
I think that it makes our training scripts a bit more flexible / easily hackable if we use the `get_xyz` approach. It also enforces a bit of structure which will be good as we expand to including community-contributed training scripts.

Having a paradigm where this is all configured as one big block of model config code leaves us too much wiggle room to end up with ugly code as we get into more complex training methods including pretraining.

(Again, just my opinion. Happy to compromise on this)

Reply:
@ianjjohnson 

> as we expand to including community-contributed training scripts.

In this context, `get_xyz` seems less powerful IMO. When the repo starts accepting training scripts, a more rigid structure might be better, something akin to Flax's [`TrainState`](https://flax.readthedocs.io/en/latest/api_reference/flax.training.html#flax.training.train_state.TrainState). However the maintenance and readability of such scripts might be worrisome. 

As far as I understand this is just a starting point and we can change the paradigms later in the case of full scale training scripts. So looking at this like any other Keras example at the moment might be the right approach IMO. 

Having said that, I am a bit of rookie with respect to all this. 😅  So please correct me if I'm wrong.

Reply:
I'd like to avoid having a really prescriptive structure like `TrainState` because we'd like these scripts to serve as an example of model training out-of-the-box with Keras+KerasCV. Adding extra structure beyond basic Python methods would obfuscate the library usage a bit.

We're still trying to work out what the best approach is, so there's no right or wrong

Reply:
> Having a paradigm where this is all configured as one big block of model config code leaves us too much wiggle room to end up with ugly code as we get into more complex training methods including pretraining.

One point in favor of `get_xyz` is that it makes it easier for us to spot small mistakes in code reviews.  I'm happy with either scenario here, lets go with Ian's approach for now.